Add ability to retry failed message transmission with a background Sidekiq job

ndaga45 opened the pull request:
This PR is a result of the discussion around [this PR](https://github.com/stitchfix/messaging/pull/26) to the stitchfix-messaging gem (which is a wrapper around Pwwka), whose solution might be general enough to go in Pwwka itself. 

It adds an `on_error: :sidekiq` option to `Pwwka::Transmitter.send_message!` to retry a failed message transmission in a Sidekiq background job. I tried to make this work exactly like the current Resque solution, so I also included a `Pwwka::Transmitter.send_message_async_sidekiq` method and a `Pwwka::SendMessageAsyncSidekiqJob` (like the corresponding Resque `Pwwka::Transmitter.send_message_async` and `Pwwka::SendMessageAsyncJob`.

In order for this to work in a dependent app, they would need to configure Pwwka to use a Sidekiq job as its async job class (instead of the default Resque job class):
```
# in config/initializers/pwwka.rb
Pwwka.configure do |config|
  config.async_job_klass = Pwwka::SendMessageAsyncSidekiqJob #this could be a custom class as well
end
```

davetron5000 commented on README.md:
```diff
@@ -129,6 +129,13 @@ Pwwka::Transmitter.send_message!(
   * `:raise` - Log the error and raise the exception received from Bunny. (default strategy)
   * `:ignore` - Log the error and return false.
   * `:resque` - Log the error and return false. Also, enqueue a job with Resque to send the message. See `send_message_async` below. **Note, this doesn't guarantee the message will actually be sent—it just guarantees an attempt is made to queue a Resque job [which could fail]**
+  * `:sidekiq` - Log the error and return false. Also, enqueue a job with Sidekiw to send the message. See `send_message_async_sidekiq` below. **Note, this doesn't guarantee the message will actually be sent—it just guarantees an attempt is made to queue a Sidekiq job [which could fail]**. To use this option, you will need to configure Pwwka to use the Sidekiq job for async messaging sending (as  ...
```
```suggestion
  * `:sidekiq` - Log the error and return false. Also, enqueue a job with Sidekiq to send the message. See `send_message_async_sidekiq` below. **Note, this doesn't guarantee the message will actually be sent—it just guarantees an attempt is made to queue a Sidekiq job [which could fail]**. To use this option, you will need to configure Pwwka to use the Sidekiq job for async messaging sending (as opposed to the default Resque job):
```

davetron5000 commented on README.md:
````diff
@@ -178,6 +185,24 @@ Pwwka.configure do |config|
+
+To enqueue a message in a background Sidekiq job, use `Transmitter.send_message_async_sidekiq`
+```ruby
+Pwwka::Transmitter.send_message_async_sidekiq(payload, routing_key, delay_by_ms: 5000) # default delay is 0
````
Should we instead overload `send_message_async` to have like `sidekiq: true`, or `queue_with: :sidekq`?

Or, maybe have a config option like `background_job_processor: ` that defaults to resque and that one could set to sidekiq, which would change the behavior of `send_message_async` but also configure the proper background job?

rosshambrick replied:
I'm not sure if this is gross or not, but maybe we could inspect the configured `async_job_klass` to know how to handle it?

davetron5000 replied:
yeah, true.  Or have the async job class adopt a protocol i.e. method that queues a job to itself?

ndaga45 replied:
I updated this PR to try out the overloading `.send_message_async` method. what do you all think?

rosshambrick commented on pwwka.gemspec:
```diff
@@ -27,6 +27,7 @@ Gem::Specification.new do |s|
   s.add_development_dependency("rspec")
   s.add_development_dependency("resque")
   s.add_development_dependency("resque-retry")
+  s.add_development_dependency("sidekiq")
```
Is there any concern/downside to including sidekiq in resque projects?  I guess this already happens for the inverse, resque being included in sidekiq projects?

davetron5000 replied:
I think it's OK, since Resque doesn't do anything just be being included.  It should stay inert if the code doens't use it.

ndaga45 replied:
I don't think so, since [the Sidekiq](https://github.com/mperham/sidekiq#sidekiq) docs say "You can have Sidekiq and Resque run side-by-side at the same time".

ndaga45 replied:
also [the Sidekiq docs](https://github.com/mperham/sidekiq#sidekiq) say "You can have Sidekiq and Resque run side-by-side at the same time." so I think it should be okay.

jbusser commented on lib/pwwka/transmitter.rb:
```diff
@@ -43,6 +44,7 @@ def initialize
     # - :ignore (aka as send_message_safely)
     # - :raise
     # - :resque -- use Resque to try to send the message later
+    # - :sidekiq -- use Sidekiq to try to send the message later
```
I am not convinced that adding an `on_error: :sidekiq` option is the right way to go here. At this level of abstraction I think we're only concerned with answering the question of how to deal with errors we encounter when publishing messages to Rabbit, and I think that generic behaviors of "ignore," "raise," or "enqueue a job" are reasonable. I'm wondering if the "resque" and "sidekiq" options are too specific. (Putting aside for a minute that we likely have to maintain `on_error: :resque` to avoid introducing breaking changes.)

The consequence of this level of specificity (`on_error: resque` and `on_error: :sidekiq`) is that we end up with nearly identical `when :resque` and `when :sidekiq` cases, and then we end up almost writing another case statement in `.send_message_async`. 

Maybe's Dave's suggestion of a `background_job_processor:` option...

davetron5000 replied:
This is a great point - in this area of the code we really do only care if we queue a job on errors. Elsewhere we can configure how that actually works.

rosshambrick replied:
Agreed.  Encapsulating that distinction into the job strategy configured in `Pwwka.configuration.async_job_klass` would be better.

ndaga45 commented:
Okay! Here's another approach. In this solution, I've added a `background_job_processor` configuration option that can be either `:resque` or `:sidekiq` (but defaults to `:resque`). That option also controls what is set as the `async_job_klass` (although that can still be overwritten by the configuration option, as before). Then, I added an `on_error: :retry_async` (not committed to that name...) to `Pwwka::Transmitter.send_message!`. This calls `Pwwka::Transmitter.send_message_async`, which will use whichever background job processor has been configured to retry sending the message. So, if an app uses `send_message_async` with no additional configurations, it should work as it did before. However, if an app configures Pwwka to use Sidekiq, `send_message_async` enqueue a Sidekiq job. Let me know if that makes sense and what you think of this approach! I think its what we were discussing in the comments.

cc: @davetron5000 @jbusser @rosshambrick

ndaga45 commented on lib/pwwka/configuration.rb:
```diff
@@ -137,5 +143,14 @@ def omit_payload_from_log?(level_of_message_with_payload)
+
+    private
+
+    def background_jobs
```
I tried to make this a constant at the top of this file, but `Pwwka::SendMessageAsyncJob` has a dependence on `Pwwka::Configuration` so there was a weird circular dependency thing going on.

ndaga45 commented on lib/pwwka/transmitter.rb:
```diff
@@ -77,24 +79,40 @@ def self.send_message!(payload, routing_key,
             raise e
           end
+
+        when :retry_async
```
This is nearly identical to the `when` block above it. Is there a way to DRY this up? I don't think Ruby's case/when falls through.

davetron5000 replied:
TIL you can use a comma for multiple options, so

```
when :resque, :retry_async
  # ...
```

will work if they are identical.

I guess another option is to leave the duplication and add a deprecation notice to the resque block (though that will probably destroy our logs with warnings…so maybe not :)

davetron5000 commented:
This looks great to me!  Exactly what I was thinking how it'd look. 🏆

rosshambrick commented on README.md:
```diff
@@ -128,7 +128,14 @@ Pwwka::Transmitter.send_message!(
   * `:raise` - Log the error and raise the exception received from Bunny. (default strategy)
   * `:ignore` - Log the error and return false.
-  * `:resque` - Log the error and return false. Also, enqueue a job with Resque to send the message. See `send_message_async` below. **Note, this doesn't guarantee the message will actually be sent—it just guarantees an attempt is made to queue a Resque job [which could fail]**
+  * `:resque` - Log the error and return false. Also, enqueue a job with Resque to send the message. See `send_message_async` below. **Note, this doesn't guarantee the message will actually be sent—it just guarantees an attempt is made to queue a Resque job [which could fail]**. This is a legacy option and `:retry_async` should be used instead.
```
I'd be in favor of removing this options from the README altogether so it only has guidance for the current api

rosshambrick commented on lib/pwwka/configuration.rb:
```diff
@@ -10,11 +11,12 @@ class Configuration
     attr_accessor :default_prefetch
     attr_reader   :requeue_on_error
     attr_writer   :app_id
+    attr_writer   :async_job_klass
```
just as fyi since i looked it up, the checkout project seems to be the only one using this option: https://github.com/stitchfix/checkout/blob/79218a299265d539a0f867b09ee0502fe0f4c01f/config/initializers/pwwka.rb#L10

rosshambrick commented on lib/pwwka/transmitter.rb:
```diff
@@ -77,24 +79,40 @@ def self.send_message!(payload, routing_key,
+      elsif background_job_processor == :sidekiq
+        options = { delay_by_ms: delay_by_ms, type: type, message_id: message_id, headers: headers }
+        job.perform_async(payload, routing_key, options)
       end
```
i think this is perfectly fine.  another option though to clean it up could be to add an `enqueue` method to the `background_job_processor` that knows how to enqueue it's own type of job

rosshambrick commented:
One suggestion but this looks great!

rosshambrick commented:
As a reminder too, a good place to verify this is working as expected is in the rails-rabbit-test project
